azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/UserPrincipalManager.java

```diff
@@ -7,37 +7,33 @@
 
 import com.nimbusds.jose.JOSEException;
 import com.nimbusds.jose.JWSAlgorithm;
+import com.nimbusds.jose.JWSHeader;
 import com.nimbusds.jose.JWSObject;
 import com.nimbusds.jose.jwk.source.JWKSource;
-import com.nimbusds.jose.jwk.source.RemoteJWKSet;
 import com.nimbusds.jose.proc.BadJOSEException;
 import com.nimbusds.jose.proc.JWSKeySelector;
 import com.nimbusds.jose.proc.JWSVerificationKeySelector;
 import com.nimbusds.jose.proc.SecurityContext;
-import com.nimbusds.jose.util.ResourceRetriever;
 import com.nimbusds.jwt.JWTClaimsSet;
 import com.nimbusds.jwt.proc.*;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
-import java.net.MalformedURLException;
-import java.net.URL;
+import java.nio.charset.StandardCharsets;
 import java.text.ParseException;
+import java.util.Base64;
 
 public class UserPrincipalManager {
-    private static final Logger LOG = LoggerFactory.getLogger(UserPrincipalManager.class);
 
-    private final ServiceEndpoints serviceEndpoints;
-    private final ConfigurableJWTProcessor<SecurityContext> validator;
-    private final ResourceRetriever resourceRetriever;
+    private final JWKSource<SecurityContext> keySource;
 
-    public UserPrincipalManager(ServiceEndpoints serviceEndpoints, ResourceRetriever resourceRetriever) {
-        this.serviceEndpoints = serviceEndpoints;
-        this.validator = getAadJwtTokenValidator();
-        this.resourceRetriever = resourceRetriever;
+    public UserPrincipalManager(JWKSource<SecurityContext> keySource) {
+        this.keySource
+                = keySource;
     }
 
     public UserPrincipal buildUserPrincipal(String idToken) throws ParseException, JOSEException, BadJOSEException {
+        final JWSHeader jwsHeader = JWSHeader.parse(new String(
```

Comment:
As line 40, `jwsObject` will be parsed, you can parse the id_token first, and then reference the algo like `jwsObject.getHeader().getAlgoritm()`.

So the whole process may looks like below, `getAADB2CKeySource()` will create the keySource (As I mentioned below 
 to keep this uri as constructor parameter) and then you may avoiding parse the header twice.
```java
private final ConfigurableJWTProcessor<SecurityContext> processor = new DefaultJWTProcessor<>();
final JWSObject jwsObject = JWSObject.parse(idToken);
final JWSAlgorithm jwsAlgorithm = jwsObject.getHeader().getAlgorithm();

processor.setJWSKeySelector(new JWSVerificationKeySelector<>(jwsAlgorithm, getAADB2CKeySource()));

final JWTClaimsSet claimsSet = processor.process(idToken, null);

processor.getJWTClaimsSetVerifier().verify(claimsSet, null);
```

Reply:
Updated to move the parsing up, remove parsing of header.